### PR TITLE
Refactor the transport with different protocol

### PR DESCRIPTION
**Pull request by yanmxa**
Signed-off-by: myan <myan@redhat.com>

ref: https://issues.redhat.com/browse/ACM-8793

We define the transporter interface as the source of truth for different implementations. 

```go
// init the transport with different implementation/protocol: secret, strimzi operator or plain deployment
type Transporter interface {
	// define the user for each hub clusters, user lifecycle is managed by the transporter
        GenerateUserName(clusterIdentity string) string                       // "kafka-user-clusterIdentity"
	CreateUser(name string) error
	DeleteUser(name string) error

	// define the topic for each hub clusters(or shared topic), topic lifecycle is managed by the transporter
	GenerateClusterTopic(clusterIdentity string) *ClusterTopic        // (spec/status/event)
	CreateClusterTopic(topic *ClusterTopic ) error    
	DeleteClusterTopic(topic *ClusterTopic ) error

	// get the connection credential by user
	GetConnCredential(userName string) (*ConnCredential, error)
}
```

### CODE REVIEW

File: operator/pkg/controllers/hubofhubs/globalhub_middleware.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		}, r.Client)
+	}
+
+	// create the user to connect the transport instance
+	err = trans.CreateUser(transportprotocol.DefaultGlobalHubKafkaUser)
+	if err != nil {
+		return nil, err
+	}
+	// create global hub topics
+	topics := trans.GenerateClusterTopic("")
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_middleware.go:120]: Why we need to create topics here?

File: operator/pkg/controllers/hubofhubs/globalhub_middleware.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		}
+	case transport.SecretTransporter:
+		trans = transportprotocol.NewSecretTransporter(ctx, types.NamespacedName{
+			Namespace: mgh.Namespace,
+			Name:      constants.GHTransportSecretName,
+		}, r.Client)
+	}
+
+	// create the user to connect the transport instance
+	err = trans.CreateUser(transportprotocol.DefaultGlobalHubKafkaUser)
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_middleware.go:115]: Do you want to treat this user as a super user?

File: pkg/transport/transporter/secret_transporter.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+	"github.com/go-logr/logr"
+	"github.com/stolostron/multicluster-global-hub/pkg/transport"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type secretTransporter struct {
```
Comment (clyang82) [inline pkg/transport/transporter/secret_transporter.go:16]: it is not a type of transporter. if you want to define a type, maybe `BYOTransporter` make more sense.

File: operator/pkg/controllers/hubofhubs/globalhub_manager.go [hunk 4]
```diff
@@ -188,3 +174,36 @@ func (r *MulticlusterGlobalHubReconciler) manipulateObj(ctx context.Context, hoh
 
 	return nil
 }
+
+type ManagerVariables struct {
```
[the remark this answers is not in the document]
Comment (yanmxa) [reply operator/pkg/controllers/hubofhubs/globalhub_manager.go:178]: The struct is also used in the test case, a common unit test error is that a field is inconsistent between the controller and the test case.

File: operator/pkg/controllers/hubofhubs/globalhub_middleware.go (hunk 2 above)
Comment (yanmxa) [reply operator/pkg/controllers/hubofhubs/globalhub_middleware.go:115]: It doesn't need to be super used. This user is just for the global hub manager to connect the transport

File: operator/pkg/controllers/hubofhubs/globalhub_middleware.go (hunk 1 above)
Comment (yanmxa) [reply operator/pkg/controllers/hubofhubs/globalhub_middleware.go:120]: It is to make sure the topic is created before the manager is deployed.

File: pkg/transport/transporter/secret_transporter.go (hunk 3 above)
Comment (yanmxa) [reply pkg/transport/transporter/secret_transporter.go]: @clyang82 Done!

**Review by clyang82 — APPROVED** [review]
/lgtm

**Pull request merged**